components/brave_shields/browser/tracking_protection_service.h

```diff
@@ -43,14 +47,51 @@ class TrackingProtectionService : public BaseLocalDataFilesObserver {
                         const base::FilePath& install_dir,
                         const std::string& manifest) override;
 
+#if BUILDFLAG(BRAVE_STP_ENABLED)
+  bool ShouldStoreState(HostContentSettingsMap* map,
+    int render_process_id, int render_frame_id, const GURL& top_origin_url,
+    const GURL& origin_url);
+
+  void SetStartingSiteForRenderFrame(GURL starting_site, int render_process_id,
+    int render_frame_id);
+  GURL GetStartingSiteForRenderFrame(int render_process_id,
+    int render_frame_id);
+  void DeleteRenderFrameKey(int render_process_id, int render_frame_id);
+  void ModifyRenderFrameKey(int old_render_process_id, int old_render_frame_id,
+    int new_render_process_id, int new_render_frame_id);
+#endif
+
+ protected:
+#if BUILDFLAG(BRAVE_STP_ENABLED)
+  void ParseStorageTrackersData();
+  struct RenderFrameIdKey {
+    RenderFrameIdKey();
+    RenderFrameIdKey(int render_process_id, int frame_routing_id);
+
+    int render_process_id;
+    int frame_routing_id;
+
+    bool operator<(const RenderFrameIdKey& other) const;
+    bool operator==(const RenderFrameIdKey& other) const;
+  };
+#endif
+
  private:
   void OnDATFileDataReady();
   std::vector<std::string> GetThirdPartyHosts(const std::string& base_host);
 
+#if BUILDFLAG(BRAVE_STP_ENABLED)
+  std::vector<std::string> first_party_storage_trackers_;
+  std::map<RenderFrameIdKey, GURL> render_frame_key_to_starting_site_url;
+  base::Lock frame_starting_site_map_lock_;
```

Comment:
Such global locks significantly affect performance (and for now we do not have a way to measure this - perftests and/or histograms). We already can see some racing Chromium bugs that affect us, but not Chromium since our threading performance is much worse (see https://github.com/brave/brave-browser/issues/2223#issuecomment-442518176)

Reply:
This is definitely a concern, but I'm not sure if there is a better way to do this. I initially had this code as part of the `BraveShieldsWebContentObserver` but moved it out to reduce the performance impact. (see: https://github.com/brave/brave-core/blob/25a24f5d7b4f6eb6ae79f3839662f7543974ebaf/components/brave_shields/browser/brave_shields_web_contents_observer.cc#L186)

Do you have any suggestions on how to improve/fix this?

Reply:
we originally discussed something like NavigationUIData which is what extensions use to pass data to the IO thread